### PR TITLE
Modules registration

### PR DESCRIPTION
**Pull request by yalosev**
#### What this PR does / why we need it

Add ability to register modules (create CR) inside the cluster
Module GV must be passed as env variable, otherwise modules will not be exported as CR

#### Does this PR introduce a user-facing change?

NONE

```release-note
feature: Add ability to register modules (create Module CR) inside the cluster
```

### CODE REVIEW

File: pkg/module_manager/module_manager.go [hunk 1]
```diff
@@ -158,7 +161,9 @@ func NewModuleManager(ctx context.Context, dirCfg DirectoryConfig, deps *ModuleM
 
 		ValuesValidator: validation.NewValuesValidator(),
 
-		modules:                     new(ModuleSet),
+		modules:           new(ModuleSet),
+		registerModulesGV: regModulesGV,
```
Comment (nabokihms) [inline pkg/module_manager/module_manager.go:165]: Maybe it is better to use `schema.GroupVersionResource` instead of a string?

File: pkg/module_manager/module_manager.go [hunk 2]
```diff
@@ -87,6 +87,9 @@ type ModuleManager struct {
 	// All known modules from specified directories ($MODULES_DIR)
 	modules *ModuleSet
 
+	// create CR with registered modules
+	registerModulesGV string
```
[the remark this answers is not in the document]
Comment (nabokihms) [reply pkg/module_manager/module_manager.go:91]: Oh, looking at the code, I get it now. It is a toggle. But why is it a string?

File: pkg/module_manager/module.go [hunk 3]
```diff
@@ -942,6 +946,59 @@ func (mm *ModuleManager) RegisterModules() error {
...
+	if mm.registerModulesGV == "" {
+		return nil
+	}
+
+	gvr, err := client.GroupVersionResource(v1alpha1.ModuleGVK.GroupVersion().String(), v1alpha1.ModuleGVK.Kind)
+	if err != nil {
+		return err
+	}
+
+	cctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
```
Comment (nabokihms) [inline pkg/module_manager/module.go:960]:
```suggestion
	cctx, cancel := context.WithTimeout(context.Background(), time.Minute)
```
It seems like timeout is too low.

File: pkg/module_manager/apis/v1alpha1/validate.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+	log "github.com/sirupsen/logrus"
+	kwhhttp "github.com/slok/kubewebhook/v2/pkg/http"
+	kwhlogrus "github.com/slok/kubewebhook/v2/pkg/log/logrus"
+	"github.com/slok/kubewebhook/v2/pkg/model"
+	kwhvalidating "github.com/slok/kubewebhook/v2/pkg/webhook/validating"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen=false
```
Comment (nabokihms) [inline pkg/module_manager/apis/v1alpha1/validate.go:15]:
```suggestion
```
Why is this required?

File: pkg/module_manager/apis/v1alpha1/validate.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+		}, nil
+	}
+
+	return &kwhvalidating.ValidatorResult{
+		Valid:   true,
+		Message: "",
+	}, nil
+})
+
+// +k8s:deepcopy-gen=false
```
Comment (nabokihms) [inline pkg/module_manager/apis/v1alpha1/validate.go:32]:
```suggestion
```
And this one?

File: pkg/module_manager/apis/v1alpha1/validate.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/slok/kubewebhook/v2/pkg/model"
+	kwhvalidating "github.com/slok/kubewebhook/v2/pkg/webhook/validating"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen=false
+var vf = kwhvalidating.ValidatorFunc(func(ctx context.Context, review *model.AdmissionReview, obj metav1.Object) (result *kwhvalidating.ValidatorResult, err error) {
+	// UserInfo groups: [system:serviceaccounts system:serviceaccounts:d8-system system:authenticated]
+	// Extra: [authentication.kubernetes.io/pod-name:[deckhouse-7cf6ddd78-grn4m]]
+	if review.UserInfo.Username != "system:serviceaccount:d8-system:deckhouse" {
```
Comment (nabokihms) [inline pkg/module_manager/apis/v1alpha1/validate.go:19]: I started thinking that maybe instead of introducing a validating webhook, it is better to raise an APIService, but this approach also has its pros and cons, e.g., not being able to control the apiVersion of Module objects.

File: pkg/module_manager/apis/v1alpha1/module.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+func (ms *ModuleStatus) GetObjectKind() schema.ObjectKind {
+	return &moduleKind{}
+}
+
+func (mk *moduleKind) SetGroupVersionKind(_ schema.GroupVersionKind) {}
+func (mk *moduleKind) GroupVersionKind() schema.GroupVersionKind {
+	return ModuleGVK
+}
+
+var ModuleGVK = schema.GroupVersionKind{Group: "deckhouse.io", Version: "v1alpha1", Kind: "Module"}
```
Comment (nabokihms) [inline pkg/module_manager/apis/v1alpha1/module.go:45]:
```suggestion
var ModuleGVK = schema.GroupVersionKind{Group: "deckhouse.io", Version: "v1alpha1", Kind: "Module"}
```
Don't think it should be a deckhouse. It would be better to pass GVR as a parameter.

File: pkg/module_manager/apis/v1alpha1/doc.go [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+// +k8s:deepcopy-gen=package
+
+//go:generate deepcopy-gen --input-dirs github.com/flant/addon-operator/pkg/module_manager/apis/v1alpha1 -O module_generated.deepcopy --go-header-file ./boilerplate.go.txt -o /tmp
```
Comment (nabokihms) [inline pkg/module_manager/apis/v1alpha1/doc.go:21]: Is this on CI? If this is a generated spec, we should probably introduce a CI job.

File: pkg/addon-operator/operator_test.go [hunk 9]
```diff
@@ -137,7 +137,7 @@ func assembleTestAddonOperator(t *testing.T, configPath string) (*AddonOperator,
 		MetricStorage:        nil,
 		HookMetricStorage:    nil,
 	}
-	op.ModuleManager = module_manager.NewModuleManager(op.ctx, dirs, &deps)
+	op.ModuleManager = module_manager.NewModuleManager(op.ctx, dirs, &deps, "")
```
Comment (nabokihms) [inline pkg/addon-operator/operator_test.go:140]: It seems like a structured config is required form ModuleManager initialization already.

File: pkg/addon-operator/operator.go [hunk 10]
```diff
@@ -2112,3 +2112,16 @@ func (op *AddonOperator) taskPhase(tsk sh_task.Task) string {
 	}
 	return ""
 }
+
+// OnFirstConvergeDone run addon-operator business logic when first converge is done and operator is ready
+func (op *AddonOperator) OnFirstConvergeDone() {
+	go func() {
+		<-op.ConvergeState.firstRunDoneC
+		time.Sleep(1 * time.Second) // wait for service to be resolved
```
Comment (nabokihms) [inline pkg/addon-operator/operator.go:2120]:
```suggestion
		time.Sleep(1 * time.Second) // wait for service to be resolved
```
Could you elaborate? I didn't get this.

File: pkg/addon-operator/bootstrap.go [hunk 11]
```diff
@@ -72,6 +73,7 @@ func (op *AddonOperator) Assemble(modulesDir string, globalHooksDir string, temp
 	shell_operator.RegisterDebugConfigRoutes(debugServer, runtimeConfig)
 	op.RegisterDebugGlobalRoutes(debugServer)
 	op.RegisterDebugModuleRoutes(debugServer)
+	op.OnFirstConvergeDone()
```
Comment (nabokihms) [inline pkg/addon-operator/bootstrap.go:76]: Why is this required? Could you please leave a comment?

File: pkg/addon-operator/admission_http_server.go [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/flant/addon-operator/pkg/module_manager/apis/v1alpha1"
+)
+
+func (op *AddonOperator) StartAdmissionServer(listenPort, certsDir string) {
+	if certsDir == "" {
+		return
```
[the remark this answers is not in the document]
Comment (yalosev) [reply pkg/addon-operator/admission_http_server.go:17]: Don't think so. If you use addon-operator as standalone module-manager and you don't want to register admission webhooks - you can just left this env variable empty and work as before.
Treat it like backward compatibility

[the remark this answers is not in the document]
Comment (yalosev) [reply pkg/addon-operator/admission_http_server.go:17]: Add one more env variable to explicitly enable admission server

File: pkg/module_manager/module_manager.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (yalosev) [reply pkg/module_manager/module_manager.go]: Removed

File: pkg/module_manager/module_manager.go (hunk 1 above)
Comment (yalosev) [reply pkg/module_manager/module_manager.go]: Not actual

File: pkg/module_manager/apis/v1alpha1/validate.go (hunk 4 above)
Comment (yalosev) [reply pkg/module_manager/apis/v1alpha1/validate.go]: Not actual

File: pkg/module_manager/apis/v1alpha1/validate.go (hunk 5 above)
Comment (yalosev) [reply pkg/module_manager/apis/v1alpha1/validate.go]: removed

File: pkg/module_manager/apis/v1alpha1/validate.go (hunk 6 above)
Comment (yalosev) [reply pkg/module_manager/apis/v1alpha1/validate.go]: I fully agree but it seems a bit raw at the moment. I think we can discuss this idea later, in the another iteration

File: pkg/module_manager/apis/v1alpha1/module.go (hunk 7 above)
Comment (yalosev) [reply pkg/module_manager/apis/v1alpha1/module.go]: Removed to the Dechkouse

File: pkg/addon-operator/operator.go (hunk 10 above)
Comment (yalosev) [reply pkg/addon-operator/operator.go:2130]: Add comment

File: pkg/addon-operator/bootstrap.go (hunk 11 above)
Comment (yalosev) [reply pkg/addon-operator/bootstrap.go:78]: Add comment

File: pkg/module_manager/module.go [hunk 13]
```diff
@@ -979,3 +1041,15 @@ func dumpData(filePath string, data []byte) error {
...
+type ModuleProducer interface {
+	GetGVK() schema.GroupVersionKind
+	NewModule() ModuleObject
+}
+
+// ModuleObject Module CR
+type ModuleObject interface {
+	SetName(name string)
+	SetWeight(weight int)
+}
```
Comment (nabokihms) [inline pkg/module_manager/module.go:1055]: Did I get it right that there is only an interface in the addon-operator, but the implementation of the interface is the duty of third-party users?

**Review by nabokihms — APPROVED** [review]
Looks good enough.

File: pkg/module_manager/module.go (hunk 13 above)
Comment (yalosev) [reply pkg/module_manager/module.go:1055]: Yes, you got it

Comment (yalosev) [reply pkg/module_manager/module.go:1055]: You can find an implementation [here](https://github.com/deckhouse/deckhouse/blob/154238fab9ad06321bd65b36f0abf04e67ae4c27/modules/002-deckhouse/hooks/pkg/apis/v1alpha1/module.go)
